### ISSUE TITLE
Support for Azure SQL AccessToken

### ISSUE THREAD

**Issue by huesie** [drwatson1/dbup-cli#16]
Azure SQL supports AccessToken-based login which is convenient for pipeline database deployments (which DbUp-cli is already awesome for - thanks @drwatson1 !).  I.e. instead of using a SQL username and password in the connection string, an access token can be retrieved by ```az account get-access-token --resource https://database.windows.net -o tsv --query '[accessToken]'``` and used for running database upgrade.

When using an access token then ```User ID``` and ```Password``` should _not_ be passed in the connection string and the permissions are granted by the Azure Principal running the pipeline, negating the need to hold a username and password - the pipeline can just use its own permissions.

AccessToken is also one way of granting permissions to AAD-based users and groups, e.g. ```CREATE USER [App1Users] FROM EXTERNAL PROVIDER;``` - attempting to do so logged-in as a SQL User gives an error such as the following: ```Msg 33159, Level 16, State 1, Line 1
Principal 'App1Users' could not be created. Only connections established with Active Directory accounts can create other Active Directory users.```

Using an access token to connect using ```SqlConnection``` is accomplished by setting the [SqlConnection.AccessToken⇗](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlconnection.accesstoken) property and DbUp supports this property in the [AzureSqlConnectionManager](https://github.com/DbUp/DbUp/blame/b134a35eefaa349aa7c4b00c8141e46750b927c9/src/dbup-sqlserver/AzureSqlConnectionManager.cs) but seems to support retrieving the AccessToken on-the-fly but I'm proposing allowing the AccessToken to be set instead, e.g. as something like ```AzureSqlAccessToken``` additional property set in the config file?

I am happy to make the code changes, but would appreciate some feedback from @drwatson1, who will ultimately need to approve this PR.  i.e. will it even be possible to work in the way I propose due to the limitations of DbUp or will we need to support a boolean that should direct DbUp-cli to use ```AzureSqlConnectionManager``` and its retrieval of the AccessToken - some design guidance would be appreciated from @drwatson1.

**Comment by drwatson1** [comment]
There are several ways as I can understand.

1. Use `useAzureSqlIntegratedSequrity` parameter of the `DeployChanges.To.SqlDatabase` extension. It should work as talks [here](https://github.com/johanclasson/vso-agent-tasks/issues/71).

```
DeployChanges.To
        .SqlDatabase("Server=my-serves.database.windows.net; Database=MyDb;", null, true)
```

This can be done without any code changes.

2. Use AzureSqlConnectionManager. In this case, we should add a new db [provider](https://github.com/drwatson1/dbup-cli/blob/f9cb15b8e89c285a1a2c0022f67c7a001be96942/src/dbup-cli/ConfigFile/Provider.cs) 'azuresql' instead of 'sqlserver' and use automatic token retrieval. 

This way needs some code changes.

3. Using SqlConnection.AccessToken. It could solve the problem but it is not accessible directly. SqlConnection is created in the ctor of the SqlConnectionManager and wrapped into the DelegateConnectionFactory so we can't change it. I searched over DbUp repo and found many discussions and issues related to Azure SQL and it seems that DbUp does not support modern authentication methods of Azure. 

Someone suggests using DbUp fork [DbUpReboot](https://github.com/chriswill/DbUpReboot), but I think this is not appropriate in this case. 

So, I recommend you to try solutions No.1 and No.2 above. If no way works for you we will think about how to implement the 3rd solution.

In this case, I think we should fork DbUp and upgrade System.Data.SqlClient to Microsoft.Data.SqClient and implement all things that we need. For example, we can add the parameter `AccessToken` to the config file and use it as you suggested.

**Comment by huesie** [comment]
Thanks for your considered reply @drwatson1 and for the great links.

Could you please consider supporting ```option 1.``` with a new patch release of DbUp-CLI by updating [```Deploy.To.SqlDatabase⇗```](https://github.com/drwatson1/dbup-cli/blob/42bf69df14592b096d3b69b6fc3adbdb59c90194/src/dbup-cli/ConfigurationHelper.cs#L22) to code similar to the following:

```C#
var useAzureSqlIntegratedSecurity = !(connectionString.Contains("Password=") ||
                                      connectionString.Contains("Integrated Security=") ||
                                      connectionString.Contains("Trusted_Connection="));
var upgrader =
    DeployChanges.To
        .SqlDatabase(connectionString, null, useAzureSqlIntegratedSecurity)
```
from: https://github.com/johanclasson/vso-agent-tasks/issues/71#issuecomment-984650497

_Caveat_: I appreciate this overload is deprecated soon after current stable DbUp 4.5.0 but is then the same approach used for the linked VSO tools that uses DbUp stable directly - and DbUp-CLI may rely on DbUp 4.5.0 for a long time due to cautious release cycle of DbUp (not a bad thing).

DbUp-CLI would then work with the approach of setting the env var ```AzureServicesAuthConnectionString``` to value e.g. ```RunAs=Developer; DeveloperTool=AzureCli```, which is my specific use case where AzureCLI is logged-in and set to the correct subscription.

**(option 2.) New AzureSql provider**: I think is good longer-term approach but may depend on DbUp vNext for the Azure SQL connection manager not yet in DbUp's stable nuget release?  Adding new DbUp-CLI provider to support ```useAzureSqlIntegratedSecurity=true``` overload and then switch to use new AzureSql DbUp connection manager when it is available in DbUp stable - is it worth it? (code change above seems low risk and may give odd error for badly-written connection string but then that gives odd errors anyways!)

**{option 3.} ```.AccessToken``` property**: I feel this is an extra luxury that I don't need for my specific use case, so I am happy with ```Option 1.``` that allows a lot of options for managed identity and for inheriting from current AzureCLI session is sufficient for me.

**[DbUpReboot](https://github.com/chriswill/DbUpReboot)**: does _not_ support PostgreSQL, so one of the compelling reasons to switch could be to support additional databases, but in this case would seem to drop support for PostgreSQL!

**Comment by drwatson1** [comment]
I don't think that using any keywords such as "Integrated Security" or "Trusted_Connection" in a connection string is a good idea because the same keywords can be a part of a standard MS SQL Server connection string and so it breaks SQL Server compatibility when we use, for example, its integrated security mode.

For this reason, using a new `azuresql` provider is a more explicit, compatible, and safe way, I think. As an alternative, we can, of cause, use some custom environment variable as a hint, something like a `UseAzureSqlIntegratedSecutrity=True` but this mostly looks like a hack :)

So, to be clear, I suggest the following solution:

1. Add a new `azuresql` provider
2. Use the heuristic above to use the overload for that provider only:

```csharp
var useAzureSqlIntegratedSecurity = !(connectionString.Contains("Password=") ||
                                      connectionString.Contains("Integrated Security=") ||
                                      connectionString.Contains("Trusted_Connection="));
var upgrader =
    DeployChanges.To
        .SqlDatabase(connectionString, null, useAzureSqlIntegratedSecurity)
```

If the solution looks good for you I can implement it tomorrow.

**Comment by huesie** [comment]
+1 adding an azuresql provider is cleaner and explicit and I already like the look of your PR, which appeared as I was typing this!

**Comment by drwatson1** [comment]
I've just published a new version. Can you test it, please? And if it works as expected I'll close the issue.

**Comment by huesie** [comment]
Thanks @drwatson1 - it mostly works - here's the details of my testing:

## Prep

1. Create a DbUp-CLI ```.env``` test file:
```
AzureServicesAuthConnectionString=RunAs=Developer;DeveloperTool=AzureCli
CONNSTR=Server=tcp:MYAZURESQLSERVER.database.windows.net,1433;Database=MYDB;Encrypt=true;Connection Timeout=30;
```

2. Create a DbUp-CLI ```dbup.yml``` test file:
```
dbUp:
  version: 1
  provider: AzureSql  # <-NEW PROVIDER YAY!
  connectionString: $CONNSTR$
```

3. **AzureCLI** is logged-in and the correct subscription selected - confirmed with:
```
az account show -o table
az account get-access-token --resource https://database.windows.net
```

## ```Upgrade``` correctly configured => WORKS

4. Build + Run a basic DbUp-CLI upgrade: (in the ```...\src\dbup-cli\bin\Debug\net5.0\``` folder in branch ```feature/issue16```)
```
./dbup-cli.exe upgrade --verbosity detail
```
**RESULT => Worked perfectly first time!**

## ```Upgrade``` badly configured => fails as expected

5. It seemed too easy - so let's edit .env file and comment-out the extra env var:
```
#AzureServicesAuthConnectionString=RunAs=Developer;DeveloperTool=AzureCli
CONNSTR=Server=tcp:MYAZURESQLSERVER.database.windows.net,1433;Database=MYDB;Encrypt=true;Connection Timeout=30;
```

6. Re-run DbUp-CLI upgrade:
```
./dbup-cli.exe upgrade --verbosity detail
```
**Now fails, as expected with:**
```
Login failed for user '<token-identified principal>'. The server is not currently configured to accept this token.
```

## Upgrade ```--ensure``` correctly configured => FAILS

7. Restore the .env file as ```step 1.```:
```
AzureServicesAuthConnectionString=RunAs=Developer;DeveloperTool=AzureCli
CONNSTR=Server=tcp:MYAZURESQLSERVER.database.windows.net,1433;Database=MYDB;Encrypt=true;Connection Timeout=30;
```

8. Re-try DbUp-CLI upgrade **and** with the ```--ensure``` switch:
```
./dbup-cli.exe upgrade --verbosity detail  # <=still works, as before, and as expected.

./dbup-cli.exe upgrade --ensure --verbosity detail  # <=FAILS WITH ERROR:
```
The ```--ensure``` switch causes the command to fail:
```
Master ConnectionString => Data Source=tcp:--REDACTED--.database.windows.net,1433;Initial
Catalog=master;Password=;Connect Timeout=30;Encrypt=True
Could not connect to the database directly
Login failed for user ''.
```
**^ Connection string is trying to connect to ```master``` database + ```Password=``` added?**
**NB: this _may_ work if ```Password=``` was not added?** (as ```--ensure``` has worked previously with sql username + password for AzureSQL)

## ```Status``` correctly configured => WORKS (and fails as expected when badly configured)

9. Try DbUp-CLI ```status``` verb:
```
./dbup-cli.exe status --show-executed --show-not-executed
```
**RESULT => Worked perfectly:**
```
Database is up-to-date. Upgrade is not required.
```

10. Edit .env file and comment-out the extra env var and retry ```status``` verb:
**Now fails, as expected with:**
```
Login failed for user '<token-identified principal>'. The server is not currently configured to accept this token.
```

## ```Drop``` correctly configured => FAILS

11. Restore the .env file as ```step 1.``` and try ```drop``` verb:
```
./dbup-cli.exe drop
```
The ```drop``` verb causes the command to fail:
```
Master ConnectionString => Data Source=tcp:--REDACTED--.database.windows.net,1433;Initial
Catalog=master;Password=;Connect Timeout=30;Encrypt=True
Could not connect to the database directly
Login failed for user ''.
```
**^ Connection string is trying to connect to ```master``` database + ```Password=``` added?**
**NB: this _may_ work if ```Password=``` was not added?** (as ```drop``` has worked previously with sql username + password for AzureSQL)


## Summary of two issues found

```step 8.``` dbup upgrade ```--ensure``` is adding a ```Password=``` into connection string causing command to fail.
```step 11.``` dbup ```drop``` is adding a ```Password=``` into connection string causing command to fail.

**Comment by drwatson1** [comment]
There are some difficulties in implementing these two operations because they are not implemented now in the DbUp and there is no easy way to extend the library.

So I copied&pasted :) some code and modified it get access token just as it does the original code. Of course, it has many limitations and drawbacks, but I think it should work and can be changed when the DbUp implements these features.

I have some troubles using AzureSQL. Can you please retest steps 8 & 11? They should work but I'm not 100% sure.

**Comment by huesie** [comment]
Thanks @drwatson1 - I pulled and built ```feature/issue16``` branch but for me ```step 4.``` (```Upgrade``` correctly configured) is now _not_ working:
```
Login failed for user '<token-identified principal>'.
```
(NB: I am using the same AzureSQL server and database as I was using yesterday with AAD-integration enabled for AzureSQL, so I believe it _should_ still work **but this could easily be my user error** - but AzureCLI context appears correct and I am able to run ```az get-access-token``` to get a token - I will re-visit this tomorrow and have a look at the new code under debug.

**One other possibility for AzureSql provider is that ```--ensure``` and ```drop``` throw ```NotImplementedException``` and are not implemented to keep it simple for this interim release**?  As the first commit delivered 100% of the functionality I personally care about for all our Azure Pipeline deployments, so is awesome (and keeps the code changes small and simple).

(i.e. we don't really use ```--ensure``` as we create the database with AzureCLI/Terraform (otherwise there is no control over database config: SKU, backup, AAD-integration, etc.) and we also do _not_ use ```drop``` as that can be a bit unpredictable with Azure and not that useful, as we then still need to run IaC release to re-create the DB (we use a separate 'drop' DbUp in sub-directory instead that drops the database objects in a 'run always' script but leaves the database intact.  Sorry, I should maybe had mentioned all of this but I didn't appreciate that ```--ensure``` and ```drop``` are so tricky, but isn't that so often the way in I.T.!).

**Comment by drwatson1** [comment]
Thank you for your testing, it helped a lot! 

I've created a test db and reproduced your error in dropping the db. 

DROP database consists of two statements: "ALTER DATABASE [{databaseName}] SET SINGLE_USER WITH ROLLBACK IMMEDIATE;" and "DROP DATABASE [{databaseName}];". The first of them always fails, but the second one is successful all the time. The problem is in an availability group as it stands in the error message, but it seems that Azure SQL is not required to use single-user mode to drop databases. So I just deleted ALTER DATABASE operation to remove the confusing error message and all works fine as I can see.
